proposals/cli-jar-processing.md

```diff
@@ -0,0 +1,48 @@
+# Proposal : Jib CLI Jar Processing 
+
+Relevant Issue: [#2796](https://github.com/GoogleContainerTools/jib/issues/2796)
+
+# Motivation 
+Allow users to containerize arbitrary jar files without having to integrate Jib into java build tools (maven and gradle) or use a `.yaml` jib cli buildfile.
+
+# Proposal Changes
+
+## Standard Jar 
+A standard jar can be containerized in two modes, exploded or packaged. 
+
+### Exploded Mode
+Achieved by calling `jib jar ${JAR_NAME}.jar --to ${TARGET_REGISTRY}`
+The default mode for containerizing a jar. It will explode a jar into the following layers:  
+- Dependencies Layer: Contains dependencies derived from `Class-Path` in jar manifest. Note that this layer will not be created if `Class-Path` is not present in the manifest.
+- Resources Layer: Contains resources parsed from jar file. Note that it will also include `MANIFEST.MF`.
+- Classes Layer: Contains classes parsed from jar file. 
+
+**Entrypoint** : `java -cp /app/dependencies/ /app/explodedJar/ ${MAIN_CLASS}`
+
+### Packaged Mode
+Achieved by calling `jib jar ${JAR_NAME}.jar --to ${TARGET_REGISTRY} --mode packaged`.
+It will result in the following layers on the container:
+- Dependencies Layer: Contains the dependencies derived from `Class-Path` in jar manifest. Note that this layer will not be created if `Class-Path` is not present in the manifest.
+- Jar Layer: Contains the original jar.
+
+**Entrypoint** : `java -jar ${JAR_NAME}.jar`
+
+## Spring-Boot Fat Jar
+A Spring-Boot Fat Jar can be containerized in two modes, exploded or packaged. 
+
+### Exploded Mode
+Achieved by calling `jib jar ${JAR_NAME}.jar --to ${TARGET_REGISTRY}`
+The default mode for containerizing a jar. It will explode a jar according to what is specified in the `layers.idx` file of the jar, if present, or according to following format:
+- Dependencies Layer: For a dependency whose version does not contain `SNAPSHOT`.
+- Spring-Boot-Loader Layer: Contains jar loader classes.
+- Snapshot-Dependencies Layer: For a dependency whose version contains `SNAPSHOT`.
```

Comment:
We can also create a snapshot dep layer in the non-Spring Boot case.

Reply:
That is an interesting idea. I modified the standard jar description to include a snapshot dep layer. Will create an issue for this.

Reply:
Actually now that I'm looking at this again, what is the advantage of having a separate snapshot dependencies layer?

Reply:
`SNAPSHOT` layers are generally those that are created from a non-release build of a project (like nightly). They could be built locally from a project (like we do for people testing jib in our [CONTRIBUTING.md](https://github.com/GoogleContainerTools/jib/blob/master/CONTRIBUTING.md#debugging-the-jib-maven-plugin-jib-maven-plugin)), or they could be something like a nightly build from a project between releases.

As a consequence snapshot dependencies are "mutable" in a sense (1.0-SNAPSHOT could be built every day with different results), they can change between builds. If we have frequently changing dependencies in the regular dependencies layer then that layer changes on every build and we have to push all dependencies up.

If we separate out snapshot dependencies, then we can isolate them from the more stable other regular dependencies and only rebuild/repush a snapshot layer.